platform/macos/joypad_macos.mm

```diff
@@ -337,257 +309,100 @@ - (void)addMacOSJoypad:(GCController *)controller {
-					if (element == xboxGamepad.paddleButton1) {
-						Input::get_singleton()->joy_button(joy_id, JoyButton::PADDLE1,
-								xboxGamepad.paddleButton1.isPressed);
-					} else if (element == xboxGamepad.paddleButton2) {
-						Input::get_singleton()->joy_button(joy_id, JoyButton::PADDLE2,
-								xboxGamepad.paddleButton2.isPressed);
-					} else if (element == xboxGamepad.paddleButton3) {
-						Input::get_singleton()->joy_button(joy_id, JoyButton::PADDLE3,
-								xboxGamepad.paddleButton3.isPressed);
-					} else if (element == xboxGamepad.paddleButton4) {
-						Input::get_singleton()->joy_button(joy_id, JoyButton::PADDLE4,
-								xboxGamepad.paddleButton4.isPressed);
-					}
-				}
-			}
+	p_joypad.ff_effect_timestamp = p_timestamp;
+}
 
-			if (@available(macOS 12, *)) {
-				if ([gamepad isKindOfClass:[GCXboxGamepad class]]) {
-					GCXboxGamepad *xboxGamepad = (GCXboxGamepad *)gamepad;
-					if (element == xboxGamepad.buttonShare) {
-						Input::get_singleton()->joy_button(joy_id, JoyButton::MISC1,
-								xboxGamepad.buttonShare.isPressed);
-					}
-				}
-			}
-		};
-	} else if (controller.microGamepad != nil) {
-		// Micro gamepads were added in macOS 10.11 and feature just 2 buttons and a d-pad.
-		_weakify(self);
-		_weakify(controller);
-
-		controller.microGamepad.valueChangedHandler = ^(GCMicroGamepad *gamepad, GCControllerElement *element) {
-			_strongify(self);
-			_strongify(controller);
-
-			int joy_id = [self getJoyIdForController:controller];
-
-			if (element == gamepad.buttonA) {
-				Input::get_singleton()->joy_button(joy_id, JoyButton::A,
-						gamepad.buttonA.isPressed);
-			} else if (element == gamepad.buttonX) {
-				Input::get_singleton()->joy_button(joy_id, JoyButton::X,
-						gamepad.buttonX.isPressed);
-			} else if (element == gamepad.dpad) {
-				Input::get_singleton()->joy_button(joy_id, JoyButton::DPAD_UP,
-						gamepad.dpad.up.isPressed);
-				Input::get_singleton()->joy_button(joy_id, JoyButton::DPAD_DOWN,
-						gamepad.dpad.down.isPressed);
-				Input::get_singleton()->joy_button(joy_id, JoyButton::DPAD_LEFT,
-						gamepad.dpad.left.isPressed);
-				Input::get_singleton()->joy_button(joy_id, JoyButton::DPAD_RIGHT,
-						gamepad.dpad.right.isPressed);
-			}
-		};
+void JoypadMacOS::joypad_vibration_stop(GameController &p_joypad, uint64_t p_timestamp) {
+	if (!p_joypad.force_feedback) {
+		return;
+	}
+	// If there is no active vibration players, exit.
+	if (!p_joypad.rumble_context->has_active_players()) {
+		return;
 	}
 
-	// TODO: Need to add support for controller.motion which gives us access to
-	// the orientation of the device (if supported).
-}
+	p_joypad.rumble_context->stop();
 
-@end
+	p_joypad.ff_effect_timestamp = p_timestamp;
+}
 
 void JoypadMacOS::process_joypads() {
-	if (@available(macOS 11, *)) {
-		// Process vibrations in macOS 11+.
-		NSArray *keys = [observer.connectedJoypads allKeys];
-
-		for (NSNumber *key in keys) {
```

Comment:
This main loop, called every frame, results in an allocation of an NSArray each time.